File: src/main/java/ledger/model/Balance.java
package ledger.model;

import ledger.util.MonetaryUtil;

import javax.money.MonetaryAmount;

import static ledger.util.MonetaryUtil.toDouble;

public record Balance(MonetaryAmount principal, MonetaryAmount interest, MonetaryAmount fee, MonetaryAmount excess) {
    public Balance(double principal, double interest, double fee, double excess) {
        this(MonetaryUtil.toMonetaryAmount(principal), MonetaryUtil.toMonetaryAmount(interest),
                MonetaryUtil.toMonetaryAmount(fee), MonetaryUtil.toMonetaryAmount(excess));
    }

    public MonetaryAmount get(BalanceComponent component) {
        return switch (component) {
            case PRINCIPAL -> principal;
            case INTEREST -> interest;
            case FEES -> fee;
            case EXCESS -> excess;
        };
    }

    public Balance add(Balance balance) {
        return new Balance(this.principal().add(balance.principal()), this.interest()
                .add(balance.interest()), this.fee().add(balance.fee()), this.excess().add(balance.excess()));
    }

    public Balance subtract(Balance balance) {
        return new Balance(this.principal().subtract(balance.principal()), this.interest()
                .subtract(balance.interest()), this.fee().subtract(balance.fee()), this.excess()
                .subtract(balance.excess()));
    }

    public boolean equals(Balance balance) {
        return toDouble(this.principal) == toDouble(balance.principal()) &&
                toDouble(this.interest) == toDouble(balance.interest()) &&
                toDouble(this.fee) == toDouble(balance.fee()) &&
                toDouble(this.excess) == toDouble(balance.excess());
    }

    public Balance negate() {
        return new Balance(this.principal().negate(), this.interest().negate(), this.fee().negate(), this.excess()
                .negate());
    }

    public MonetaryAmount getTotalAmount() {
        return this.principal.add(this.interest).add(this.fee).add(this.excess);
    }
}

File: haskell/src/Ledger/Balance.hs
-- | Balance arithmetic — pure functions over the four-component balance.
module Ledger.Balance
  ( addBalance
  , subtractBalance
  , negateBalance
  , totalAmount
  , getComponent
  , setComponent
  ) where

import Ledger.Types

-- | Component-wise addition of two balances.
--
-- > addBalance (Balance 100 10 5 0) (Balance 200 20 10 0)
-- >   == Balance 300 30 15 0
addBalance :: Balance -> Balance -> Balance
addBalance a b = Balance
  { balPrincipal = balPrincipal a + balPrincipal b
  , balInterest  = balInterest  a + balInterest  b
  , balFee       = balFee       a + balFee       b
  , balExcess    = balExcess    a + balExcess    b
  }

-- | Component-wise subtraction.
subtractBalance :: Balance -> Balance -> Balance
subtractBalance a b = Balance
  { balPrincipal = balPrincipal a - balPrincipal b
  , balInterest  = balInterest  a - balInterest  b
  , balFee       = balFee       a - balFee       b
  , balExcess    = balExcess    a - balExcess    b
  }

-- | Negate every component.
negateBalance :: Balance -> Balance
negateBalance b = Balance
  { balPrincipal = negate (balPrincipal b)
  , balInterest  = negate (balInterest  b)
  , balFee       = negate (balFee       b)
  , balExcess    = negate (balExcess    b)
  }

-- | Sum of all four components.
totalAmount :: Balance -> Money
totalAmount b = balPrincipal b + balInterest b + balFee b + balExcess b

-- | Project a single component from a balance.
getComponent :: BalanceComponent -> Balance -> Money
getComponent Principal b = balPrincipal b
getComponent Interest  b = balInterest  b
getComponent Fees      b = balFee       b
getComponent Excess    b = balExcess    b

-- | Return a new balance with one component replaced.
setComponent :: BalanceComponent -> Money -> Balance -> Balance
setComponent Principal v b = b { balPrincipal = v }
setComponent Interest  v b = b { balInterest  = v }
setComponent Fees      v b = b { balFee       = v }
setComponent Excess    v b = b { balExcess    = v }
